### ISSUE TITLE
PubSubEngine interface's publish function should not be synchronous

### ISSUE THREAD

**Issue by grantwwu** [apollographql/graphql-subscriptions#160]
[The PubSubEngine interface](https://github.com/apollographql/graphql-subscriptions/blob/master/src/pubsub-engine.ts#L2) specifies the publish function as:
```
publish(triggerName: string, payload: any): boolean;
```
which is synchronous.

In my opinion, this is inappropriate for any production (i.e., anything backed by an external store) deployment.  In those implementations, publishing often goes over the network, and might fail or otherwise be lost.

The current semantics of `PubSubEngine` aren't well documented, but assuming that the return value of `publish` should indicate a success or failure, they demand that implementations can synchronously determine success or failure, which isn't feasible if they go over the network.

Looking over what the current linked implementations do:
- The Redis implementation has a [TODO saying to file an issue about this](https://github.com/davidyaha/graphql-redis-subscriptions/blob/master/src/redis-pubsub.ts#L62) cc @davidyaha  
- The Google Pubsub implementation [ignores the interface typing](https://github.com/axelspringer/graphql-google-pubsub/blame/master/src/google-pubsub.js#L23) and [returns a promise](https://cloud.google.com/nodejs/docs/reference/pubsub/0.19.x/Publisher#publish) cc @jonas-arkulpa 
- The MQTT implementation creates a [floating promise](https://github.com/davidyaha/graphql-mqtt-subscriptions/blob/master/src/mqtt-pubsub.ts#L51) and [returns true unconditionally](https://github.com/davidyaha/graphql-mqtt-subscriptions/blob/master/src/mqtt-pubsub.ts#L52)
- The RabbitMQ implementation does the same thing: [floats a](https://github.com/cdmbase/graphql-rabbitmq-subscriptions/blob/master/src/amqp-pubsub.ts#L51) [promise](https://github.com/cdmbase/rabbitmq-pubsub/blob/master/src/publisher.ts#L14), [returns true](https://github.com/cdmbase/graphql-rabbitmq-subscriptions/blob/master/src/amqp-pubsub.ts#L52)
- It's a bit hard to disentangle what the [Kafka implementation does](https://github.com/ancashoria/graphql-kafka-subscriptions/blob/master/src/kafka-pubsub.ts#L50), but near as I can tell, it writes to a [NodeJS stream](https://nodejs.org/api/stream.html#stream_writable_write_chunk_encoding_callback), not supplying an data flush callback.  I think this is morally equivalent to floating a promise, but with the boolean value indicating when the stream isn't draining.
- The [Postgres implementation](https://github.com/GraphQLCollege/graphql-postgres-subscriptions/blob/master/postgres-pubsub.js#L20) uses an event emitter, which I don't think has any guaranteed delivery semantics, and returns true.
- The [NATS implementation](https://github.com/moonwalker/graphql-nats-subscriptions/blob/master/src/nats-pubsub.ts#L17) does not supply a callback to the [publish function](https://github.com/nats-io/node-nats/blob/master/lib/nats.js#L1164) and returns true.

tl:dr; Please make `publish` return a `Promise<boolean>`.

**Comment by NeoPhi** [comment]
I'd welcome a PR for this change. To make the migration smoother I'd propose `boolean | Promise<boolean>`.

**Comment by grantwwu** [comment]
On second thought - should it instead return `Promise<void>`, and with a rejected promise signaling error?

I'm not actually very experienced with Promise API design.

**Comment by grantwwu** [comment]
It makes the API a bit ambiguous, though, doesn't it?  A successful publish will always result in a settled Promise holding `true`, but a failed one could both be represented as a settled Promise holding `false` or a rejected Promise.

**Comment by grantwwu** [comment]
I mean the thing is, by using Promises, we're already opting into Promise rejection, which is a completely different and more complex API.  This sort of situation where you have a large, open-ended set of reasons why something could go wrong is what exceptions are designed for.

Look at it another way - why one would ever want to get a settled Promise holding false?  As far as I can tell, the use of exceptions for error handling is pervasive in the Javascript ecosystem.  Therefore, any consumer of this Promise based API is already going to need to handle rejected promises, so returning `Promise.resolve(false)` is actually adding a third case for consumers to handle.

**Comment by NeoPhi** [comment]
Good points. My original hope was to keep the delta small but by adding a possible Promise as a return we break backwards compatibility. So since this will be a breaking change anyway, switching to just Promise<void> makes the most sense.

**Comment by jonas-arkulpa** [comment]
@NeoPhi Do you plan to release this change in the near future?
I currently rewrite https://github.com/axelspringer/graphql-google-pubsub in TypeScript and would like to have the fixed type as I ignored the type before and just returned a promise.
